Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -136,16 +138,18 @@ class TopSitesPanel: UIViewController {
     }
 
     private func deleteHistoryTileForSite(site: Site, atIndexPath indexPath: NSIndexPath) {
-        func reloadThumbnails() {
-            self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
-                .uponQueue(dispatch_get_main_queue()) { result in
+        func reloadThumbnails() -> Success {
+            return self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
+                .bindQueue(dispatch_get_main_queue()) { result in
                     self.deleteOrUpdateSites(result, indexPath: indexPath)
```

Comment:
Your approach makes a lot more sense and simplifies the logic. One issue though is the way we're querying for Top Sites. We're requesting `maxFrecencyLimit` number of cells from the cache to display. `maxFrecencyLimit` is an approximation for the max number of items we'll see in either landscape or portrait orientation. The code currently 'blocks' on the delete until a request is made to get the new set of data from the cache before proceeding. If the user deletes really fast it's easy to get past the number of items we pull from the cache. 

It's similar to the problem of having an infinite list since we don't know how many items there are in total for the Top Sites panel. I mean, we could technically grab the whole cache but since it's calculated based on history it could be huge. Actually, the more I think about it, the more this maps to a producer/consumer problem. The user is 'consuming' items from the panel while the cache queries are 'producing' more items for the panel to display. 

Any thoughts on solving that pickle?